GCP bucket backup container

msheiny opened the pull request:
Adds a new container spec for making GCP bucket backups:

* Takes a source bucket
* Brings that down locally, tars it up
* Uploads that to another bucket with user-provided encryption (a GCP feature)

Note I didn't include a metadata file here cause I'm not building with the `build.py` scripts in the root. I'm trying something new and wacky.

To test? Well I've been testing locally using something like:
`podman run --privileged -e GBACKUP_ENV_KEY=$(cat gskey) -v "$(pwd)/gsutil.json:/secret/gcp.key:ro" -v "$(pwd)/bucket_sync.py:/usr/local/bin/bucket_sync.py" -it fpf.local/gsutil_sync:latest gs://stn-fpf-media/media/ gs://k8s-www-backups/stn stnmedia -svc_act_key=/secret/gcp.key`

A better thing to do would give this a visual glance over and give a more thorough investigation of the implementation as a kubernetes cron-job

msheiny committed:
Foundation of bucket sync script

Current functionality will sync a remote bucket to a local temp
directory. Has bare minimum of tests on one class function. Some
argparse basics. Adding this while at least one component is working :)

msheiny committed:
Get tar and encrypted upload logic in place

Could use some more tests but lets get back to that later

msheiny committed:
Add gcloud init logic, hammer out Dockerfile

msheiny committed:
Rename to better reflect image purpose

msheiny committed:
Update tests for latest class invocation params

msheiny committed:
Sets loglevel to INFO by default

Re-classifies some formerly INFO level log events -> DEBUG. Main
reasoning here is I wanted at least one log output for successful
events.

msheiny committed:
Sanitize encryption keys from being dumped to logs

Only applicable in verbose mode, will look something like:
> Calling command /usr/bin/gsutil -o "GSUtil:encryption_key=XXXXXXXXXX" cp /tmp/tmpav9t25qb/src-bucket.tar.gz gs://dest/stnmedia.tar.gz

msheiny committed:
Add basic tests

msheiny committed:
Clean-up pipenv lock env for python3

msheiny commented:
rebased on `e8de354`

conorsch commented on gs_bucket_backup/bucket_sync.py:
```diff
@@ -0,0 +1,168 @@
+        """ Given a file path upload said file to our backup bucket.
+            File is timestamp'd and includes file prefix. """
+        timestamp_name = datetime.datetime.now().strftime(
+            '%b-%d-%y-%X-{}.tar.gz'.format(self.suffix))
```
Would strongly prefer `'%F-%R-%s` for the timestamp formatting, same as we use in the db backups.

conorsch commented on gs_bucket_backup/test_bucket_sync.py:
```diff
@@ -0,0 +1,47 @@
+from bucket_sync import GCPBucketBackup
+
+
+class TestBucketSync(object):
```
Tests! Oh, my! :smiley:

conorsch replied:
Was not able to run these tests locally, even with Python 3.5.3, due to a missing `pathlib2` import. Once added, tests passed A-OK. Will tack on to the reqs.

msheiny replied:
you used `pipenv run pytest` (after `pipenv install`) ? Huh its working in CI and locally for me. I did see this issue early on when I created a local pipenv with python 2.x - let me know what you find out?

conorsch commented on gs_bucket_backup/Dockerfile:
```diff
@@ -0,0 +1,13 @@
+# tag 258.0.0
```
Nit: let's rename this container to `gke-gs-bucket-backup`. A few reasons:

- all other container names use hyphens rather than underscores
- we'll likely have a pile of GKE-related helper containers, so naming them with a prefix will aid in discovery
- in order to update the name, we'll have to touch the corresponding config references elsewhere, which will greatly aid in review here

conorsch reviewed (changes requested):
Requesting a few changes, but also offering to implement them, mostly as a scavenger hunt to track down the relevant points of update. Let me know if you disagree with any of the comments placed in-line, otherwise I'll proceed with updating.

conorsch committed:
Renames GKE bucket backup container

Two name changes:

  1. container image: "gs_bucket_backup" -> "gke-gs-bucket-backup"
  2. script: "bucket_sync.py" -> "gs_bucket_sync.py"

Using a prefix "gke-" on the container name, since we'll likely have a
handful of these helpers in the near future.

conorsch committed:
Updates timestamp format in backup filename

Before:
    Aug-22-19-12:31:41-stnmedia.tar.gz

After:
    2019-08-21-20:00-1566417638-stnmedia.tar.gz

Matches the formatting on the db backups we prepare via other means.

conorsch commented:
Working well, after putting in the corresponding changes in other repos to update the container name. Satisfied with functional review performed locally against dev cluster.

conorsch commented:
GitHub is asking for a rebase, so doing that on top of 07b4804833358b1278777f863825fa7b8ddb307a prior to merge.
